File: Android_Version/app/src/main/java/com/sepahi/ghostlark/core/SingBoxConfig.kt
package com.sepahi.ghostlark.core

import com.sepahi.ghostlark.model.AppSettings
import com.sepahi.ghostlark.model.Proto
import com.sepahi.ghostlark.model.ProxyNode
import com.sepahi.ghostlark.model.TlsMode
import com.sepahi.ghostlark.model.Transport
import com.sepahi.ghostlark.model.WarpAccount
import org.json.JSONArray
import org.json.JSONObject

/** Builds sing-box 1.14 JSON configurations. Mirrors the macOS builder, with a TUN inbound for Android. */
object SingBoxConfig {
    const val WARP_PEER_PUBLIC_KEY = "bmXOC+F1FxEMF9dyiK2H5/1SUtzH0JuVo51h2wPfgyo="

    class Options(val fragment: Boolean, val fragmentReality: Boolean, val fingerprint: String)

    private fun opts(s: AppSettings) = Options(s.stealthMode && s.tlsFragment, s.fragmentReality, s.utlsFingerprint)

    private fun jo(vararg pairs: Pair<String, Any?>): JSONObject = JSONObject().also { o -> pairs.forEach { (k, v) -> if (v != null) o.put(k, v) } }
    private fun ja(items: List<Any>): JSONArray = JSONArray().also { a -> items.forEach { a.put(it) } }

    private val knownFingerprints = setOf("chrome", "firefox", "edge", "safari", "360", "qq", "ios", "android", "random", "randomized")

    fun outbound(n: ProxyNode, tag: String, o: Options): JSONObject {
        val out = jo("tag" to tag, "server" to n.server, "server_port" to n.port)
        when (n.proto) {
            Proto.VLESS -> { out.put("type", "vless"); out.put("uuid", n.uuid ?: ""); n.flow?.takeIf { it.isNotEmpty() }?.let { out.put("flow", it) }; out.put("packet_encoding", "xudp") }
            Proto.VMESS -> { out.put("type", "vmess"); out.put("uuid", n.uuid ?: ""); out.put("security", n.vmessSecurity ?: "auto"); out.put("alter_id", n.alterId); out.put("packet_encoding", "xudp") }
            Proto.TROJAN -> { out.put("type", "trojan"); out.put("password", n.password ?: "") }
            Proto.SHADOWSOCKS -> {
                out.put("type", "shadowsocks"); out.put("method", n.method ?: "aes-256-gcm"); out.put("password", n.password ?: "")
                n.plugin?.let { out.put("plugin", it); out.put("plugin_opts", n.pluginOpts ?: "") }
            }
            Proto.HYSTERIA2 -> {
                out.put("type", "hysteria2"); out.put("password", n.password ?: "")
                n.obfsType?.let { out.put("obfs", jo("type" to it, "password" to (n.obfsPassword ?: ""))) }
            }
            Proto.TUIC -> {
                out.put("type", "tuic"); out.put("uuid", n.uuid ?: ""); out.put("password", n.password ?: "")
                out.put("congestion_control", n.congestionControl ?: "bbr"); out.put("udp_relay_mode", n.udpRelayMode ?: "native")
            }
            Proto.WIREGUARD -> { out.put("type", "direct"); return out }
        }
        val isQuic = n.proto == Proto.HYSTERIA2 || n.proto == Proto.TUIC
        if (n.security != TlsMode.NONE || isQuic) {
            val tls = jo("enabled" to true)
            n.effectiveSni?.let { tls.put("server_name", it) }
            if (n.insecure) tls.put("insecure", true)
            if (n.alpn.isNotEmpty()) tls.put("alpn", ja(n.alpn))
            if (!isQuic) {
                val fp = n.fingerprint?.lowercase()?.takeIf { it in knownFingerprints } ?: o.fingerprint
                tls.put("utls", jo("enabled" to true, "fingerprint" to fp))
            }
            if (n.security == TlsMode.REALITY && n.realityPublicKey != null) {
                val r = jo("enabled" to true, "public_key" to n.realityPublicKey)
                n.realityShortId?.takeIf { it.isNotEmpty() }?.let { r.put("short_id", it) }
                tls.put("reality", r)
            }
            if (o.fragment && !isQuic && (n.security != TlsMode.REALITY || o.fragmentReality)) {
                tls.put("fragment", true); tls.put("record_fragment", true)
            }
            out.put("tls", tls)
        }
        when (n.transport) {
            Transport.WS -> {
                val t = jo("type" to "ws")
                var path = n.path ?: "/"
                val ed = path.indexOf("?ed=")
                if (ed >= 0) {
                    val num = path.substring(ed + 4).takeWhile { it.isDigit() }.toIntOrNull() ?: 2048
                    path = path.substring(0, ed)
                    t.put("max_early_data", num); t.put("early_data_header_name", "Sec-WebSocket-Protocol")
                }
                t.put("path", path.ifEmpty { "/" })
                n.host?.takeIf { it.isNotEmpty() }?.let { t.put("headers", jo("Host" to it)) }
                out.put("transport", t)
            }
            Transport.GRPC -> out.put("transport", jo("type" to "grpc", "service_name" to (n.serviceName ?: n.path ?: "")))
            Transport.HTTP -> {
                val t = jo("type" to "http", "path" to (n.path ?: "/"))
                n.host?.takeIf { it.isNotEmpty() }?.let { t.put("host", ja(it.split(','))) }
                out.put("transport", t)
            }
            Transport.HTTPUPGRADE -> {
                val t = jo("type" to "httpupgrade", "path" to (n.path ?: "/"))
                n.host?.takeIf { it.isNotEmpty() }?.let { t.put("host", it) }
                out.put("transport", t)
            }
            else -> {}
        }
        return out
    }

    fun warpEndpoint(a: WarpAccount, tag: String, detour: String?): JSONObject {
        val e = jo(
            "type" to "wireguard", "tag" to tag, "address" to ja(listOf("${a.v4}/32", "${a.v6}/128")),
            "private_key" to a.privateKey, "mtu" to 1280,
            "peers" to ja(listOf(jo(
                "address" to a.endpointHost, "port" to a.endpointPort, "public_key" to a.peerPublicKey,
                "allowed_ips" to ja(listOf("0.0.0.0/0", "::/0")), "reserved" to ja(a.reserved),
            ))),
        )
        detour?.let { e.put("detour", it) }
        return e
    }

    fun mainConfig(node: ProxyNode, s: AppSettings, warp: WarpAccount?, apiSecret: String, backups: List<ProxyNode> = emptyList()): String {
        val o = opts(s)
        val shield = s.shieldMode && warp != null && !node.isWarp
        val exit = if (shield || node.isWarp) "warp" else "proxy"
        val outbounds = mutableListOf<Any>()
        val endpoints = mutableListOf<Any>()
        if (node.isWarp && warp != null) endpoints += warpEndpoint(warp, "warp", null)
        else {
            if (s.extraStealthOn && backups.isNotEmpty()) {
                // Failover pool: a blocked server is replaced by the core itself, never by the bare connection.
                val members = listOf(node) + backups
                members.forEach { outbounds += outbound(it, it.tag, o) }
                outbounds += jo("type" to "urltest", "tag" to "proxy", "outbounds" to ja(members.map { it.tag }),
                    "url" to s.testUrl, "interval" to "5m", "tolerance" to 150, "interrupt_exist_connections" to false)
            } else outbounds += outbound(node, "proxy", o)
            if (shield && warp != null) endpoints += warpEndpoint(warp, "warp", "proxy")
        }
        outbounds += jo("type" to "direct", "tag" to "direct")
        val dnsRules = mutableListOf<Any>()
        val routeRules = mutableListOf<Any>(
            jo("action" to "sniff"),
            jo("protocol" to "dns", "action" to "hijack-dns"),
            jo("ip_is_private" to true, "outbound" to "direct"),
        )
        if (s.extraStealthOn) routeRules.add(2, jo("network" to "udp", "port" to 443, "action" to "reject"))
        if (s.stealthMode && s.domesticDirect) {
            dnsRules += jo("domain_suffix" to ja(listOf(".ir")), "server" to "dns-local")
            routeRules += jo("domain_suffix" to ja(listOf(".ir")), "outbound" to "direct")
        }
        val cfg = jo(
            "log" to jo("level" to "info", "timestamp" to true),
            "dns" to jo(
                "servers" to ja(listOf(
                    jo("type" to "https", "tag" to "dns-remote", "server" to "1.1.1.1", "detour" to exit),
                    jo("type" to "local", "tag" to "dns-local"),
                )),
                "rules" to ja(dnsRules), "final" to "dns-remote", "strategy" to "prefer_ipv4",
            ),
            "inbounds" to ja(listOf(jo(
                "type" to "tun", "tag" to "tun-in", "address" to ja(listOf("172.19.0.1/30", "fdfe:dcba:9876::1/126")),
                "mtu" to 1500, "auto_route" to true, "strict_route" to true, "stack" to "mixed",
            ))),
            "outbounds" to ja(outbounds),
            "route" to jo("rules" to ja(routeRules), "final" to exit, "auto_detect_interface" to true, "default_domain_resolver" to "dns-local"),
            "experimental" to jo("clash_api" to jo("external_controller" to "127.0.0.1:${s.apiPort}", "secret" to apiSecret)),
        )
        if (endpoints.isNotEmpty()) cfg.put("endpoints", ja(endpoints))
        return cfg.toString(2)
    }

    /** Many outbounds, no inbound: used for batch latency tests without touching the VPN. */
    fun testerConfig(nodes: List<ProxyNode>, s: AppSettings, warp: WarpAccount?, apiSecret: String): String {
        val o = opts(s)
        val outbounds = mutableListOf<Any>()
        val endpoints = mutableListOf<Any>()
        for (n in nodes) {
            if (n.isWarp) { if (warp != null) endpoints += warpEndpoint(warp, n.tag, null) } else outbounds += outbound(n, n.tag, o)
        }
        outbounds += jo("type" to "direct", "tag" to "direct")
        val cfg = jo(
            "log" to jo("level" to "error"),
            "dns" to jo("servers" to ja(listOf(jo("type" to "local", "tag" to "dns-local"))), "final" to "dns-local", "strategy" to "prefer_ipv4"),
            "outbounds" to ja(outbounds),
            "route" to jo("final" to "direct", "auto_detect_interface" to true, "default_domain_resolver" to "dns-local"),
            "experimental" to jo("clash_api" to jo("external_controller" to "127.0.0.1:${s.testerApiPort}", "secret" to apiSecret)),
        )
        if (endpoints.isNotEmpty()) cfg.put("endpoints", ja(endpoints))
        return cfg.toString(2)
    }
}

File: Ghostlark/Sources/Core/SingBoxConfig.swift
import Foundation

/// Builds sing-box 1.14 JSON configurations from ProxyNodes and settings.
///
/// `auto_detect_interface` stays off on macOS: Ghostlark runs as a system proxy, not a TUN, so the core's own
/// connections can never loop back into it. Binding to the physical interface would only break Ghostlark when
/// another VPN (Cloudflare WARP, Proton, …) owns the default route and drops traffic that bypasses it.
enum SingBoxConfig {

    static let warpPeerPublicKey = "bmXOC+F1FxEMF9dyiK2H5/1SUtzH0JuVo51h2wPfgyo="

    struct Options {
        var stealth: Bool
        var fragment: Bool
        var fragmentReality: Bool
        var fingerprint: String
    }

    // MARK: Outbound for one node

    static func outbound(for node: ProxyNode, tag: String, opts: Options) -> [String: Any] {
        var o: [String: Any] = ["tag": tag, "server": node.server, "server_port": node.port]

        switch node.proto {
        case .vless:
            o["type"] = "vless"
            o["uuid"] = node.uuid ?? ""
            if let f = node.flow, !f.isEmpty { o["flow"] = f }
            o["packet_encoding"] = "xudp"
        case .vmess:
            o["type"] = "vmess"
            o["uuid"] = node.uuid ?? ""
            o["security"] = node.vmessSecurity ?? "auto"
            o["alter_id"] = node.alterId
            o["packet_encoding"] = "xudp"
        case .trojan:
            o["type"] = "trojan"
            o["password"] = node.password ?? ""
        case .shadowsocks:
            o["type"] = "shadowsocks"
            o["method"] = node.method ?? "aes-256-gcm"
            o["password"] = node.password ?? ""
            if let p = node.plugin { o["plugin"] = p; o["plugin_opts"] = node.pluginOpts ?? "" }
        case .hysteria2:
            o["type"] = "hysteria2"
            o["password"] = node.password ?? ""
            if let t = node.obfsType { o["obfs"] = ["type": t, "password": node.obfsPassword ?? ""] }
        case .tuic:
            o["type"] = "tuic"
            o["uuid"] = node.uuid ?? ""
            o["password"] = node.password ?? ""
            o["congestion_control"] = node.congestionControl ?? "bbr"
            o["udp_relay_mode"] = node.udpRelayMode ?? "native"
        case .wireguard:
            // WireGuard is an endpoint, not an outbound; handled by warpEndpoint().
            o["type"] = "direct"
            return o
        }

        // TLS
        let needsTLS = node.security != .none || node.proto == .hysteria2 || node.proto == .tuic
        if needsTLS {
            var tls: [String: Any] = ["enabled": true]
            if let sni = node.effectiveSNI { tls["server_name"] = sni }
            if node.insecure { tls["insecure"] = true }
            if !node.alpn.isEmpty { tls["alpn"] = node.alpn }
            let isQUIC = node.proto == .hysteria2 || node.proto == .tuic
            if !isQUIC {
                tls["utls"] = ["enabled": true, "fingerprint": normalizeFingerprint(node.fingerprint) ?? opts.fingerprint]
            }
            if node.security == .reality, let pbk = node.realityPublicKey {
                var r: [String: Any] = ["enabled": true, "public_key": pbk]
                if let sid = node.realityShortId, !sid.isEmpty { r["short_id"] = sid }
                tls["reality"] = r
            }
            // TLS ClientHello fragmentation defeats SNI-based blocking of the *proxy server itself*.
            if opts.fragment && !isQUIC && (node.security != .reality || opts.fragmentReality) {
                tls["fragment"] = true
                tls["record_fragment"] = true
            }
            o["tls"] = tls
        }

        // Transport
        switch node.transport {
        case .ws:
            var t: [String: Any] = ["type": "ws"]
            var path = node.path ?? "/"
            if let q = path.range(of: "?ed=") {
                let ed = Int(path[q.upperBound...].prefix { $0.isNumber }) ?? 2048
                path = String(path[..<q.lowerBound])
                t["max_early_data"] = ed
                t["early_data_header_name"] = "Sec-WebSocket-Protocol"
            }
            t["path"] = path.isEmpty ? "/" : path
            if let h = node.host, !h.isEmpty { t["headers"] = ["Host": h] }
            o["transport"] = t
        case .grpc:
            o["transport"] = ["type": "grpc", "service_name": node.serviceName ?? node.path ?? ""]
        case .http:
            var t: [String: Any] = ["type": "http", "path": node.path ?? "/"]
            if let h = node.host, !h.isEmpty { t["host"] = h.split(separator: ",").map { String($0) } }
            o["transport"] = t
        case .httpupgrade:
            var t: [String: Any] = ["type": "httpupgrade", "path": node.path ?? "/"]
            if let h = node.host, !h.isEmpty { t["host"] = h }
            o["transport"] = t
        case .tcp, .xhttp, .kcp, .quic:
            break
        }
        return o
    }

    static func normalizeFingerprint(_ fp: String?) -> String? {
        guard let fp = fp?.lowercased(), !fp.isEmpty else { return nil }
        let known = ["chrome", "firefox", "edge", "safari", "360", "qq", "ios", "android", "random", "randomized"]
        return known.contains(fp) ? fp : nil
    }

    static func warpEndpoint(account: WARPAccount, tag: String, detour: String?) -> [String: Any] {
        var e: [String: Any] = [
            "type": "wireguard",
            "tag": tag,
            "address": ["\(account.v4)/32", "\(account.v6)/128"],
            "private_key": account.privateKey,
            "mtu": 1280,
            "peers": [[
                "address": account.endpointHost,
                "port": account.endpointPort,
                "public_key": account.peerPublicKey,
                "allowed_ips": ["0.0.0.0/0", "::/0"],
                "reserved": account.reserved,
            ] as [String: Any]],
        ]
        if let d = detour { e["detour"] = d }
        return e
    }

    // MARK: Main (connection) config

    static func mainConfig(node: ProxyNode, backups: [ProxyNode] = [], settings: AppSettings, warp: WARPAccount?, apiSecret: String) -> [String: Any] {
        let opts = Options(stealth: settings.stealthMode, fragment: settings.stealthMode && settings.tlsFragment,
                           fragmentReality: settings.fragmentReality, fingerprint: settings.utlsFingerprint)
        let shield = settings.shieldMode && warp != nil && !node.isWARP
        let exitTag = (shield || node.isWARP) ? "warp" : "proxy"

        var outbounds: [[String: Any]] = []
        var endpoints: [[String: Any]] = []
        if node.isWARP, let w = warp {
            endpoints.append(warpEndpoint(account: w, tag: "warp", detour: nil))
        } else {
            if settings.extraStealthOn && !backups.isEmpty {
                // Failover pool: if the active server is blocked mid-session the core moves to the next
                // verified one by itself, so traffic never falls back to the bare connection.
                let members = [node] + backups
                for m in members { outbounds.append(outbound(for: m, tag: m.tag, opts: opts)) }
                outbounds.append([
                    "type": "urltest", "tag": "proxy", "outbounds": members.map { $0.tag },
                    "url": settings.testURL, "interval": "5m", "tolerance": 150, "interrupt_exist_connections": false,
                ])
            } else {
                outbounds.append(outbound(for: node, tag: "proxy", opts: opts))
            }
            if shield, let w = warp {
                endpoints.append(warpEndpoint(account: w, tag: "warp", detour: "proxy"))
            }
        }
        outbounds.append(["type": "direct", "tag": "direct"])

        var dnsRules: [[String: Any]] = []
        var routeRules: [[String: Any]] = [
            ["action": "sniff"],
            ["protocol": "dns", "action": "hijack-dns"],
            ["ip_is_private": true, "outbound": "direct"],
        ]
        if settings.extraStealthOn {
            // QUIC relayed through a TCP proxy is a distinctive pattern and most free servers drop UDP anyway;
            // rejecting it makes browsers fall back to ordinary HTTPS over TCP at once.
            routeRules.insert(["network": "udp", "port": 443, "action": "reject"], at: 2)
        }
        if settings.stealthMode && settings.domesticDirect {
            dnsRules.append(["domain_suffix": [".ir"], "server": "dns-local"])
            routeRules.append(["domain_suffix": [".ir"], "outbound": "direct"])
        }

        var config: [String: Any] = [
            "log": ["level": "info", "timestamp": true],
            "dns": [
                "servers": [
                    ["type": "https", "tag": "dns-remote", "server": "1.1.1.1", "detour": exitTag],
                    ["type": "local", "tag": "dns-local"],
                ],
                "rules": dnsRules,
                "final": "dns-remote",
                "strategy": "prefer_ipv4",
            ] as [String: Any],
            "inbounds": [[
                "type": "mixed", "tag": "mixed-in", "listen": "127.0.0.1", "listen_port": settings.localPort,
            ] as [String: Any]],
            "outbounds": outbounds,
            "route": [
                "rules": routeRules,
                "final": exitTag,
                "auto_detect_interface": false,
                "default_domain_resolver": "dns-local",
            ] as [String: Any],
            "experimental": [
                "clash_api": ["external_controller": "127.0.0.1:\(settings.apiPort)", "secret": apiSecret],
            ],
        ]
        if !endpoints.isEmpty { config["endpoints"] = endpoints }
        return config
    }

    // MARK: Tester config (many outbounds, no inbound)

    static func testerConfig(nodes: [ProxyNode], settings: AppSettings, warp: WARPAccount?, apiSecret: String) -> [String: Any] {
        let opts = Options(stealth: settings.stealthMode, fragment: settings.stealthMode && settings.tlsFragment,
                           fragmentReality: settings.fragmentReality, fingerprint: settings.utlsFingerprint)
        var outbounds: [[String: Any]] = []
        var endpoints: [[String: Any]] = []
        for n in nodes {
            if n.isWARP {
                if let w = warp { endpoints.append(warpEndpoint(account: w, tag: n.tag, detour: nil)) }
            } else {
                outbounds.append(outbound(for: n, tag: n.tag, opts: opts))
            }
        }
        outbounds.append(["type": "direct", "tag": "direct"])
        var config: [String: Any] = [
            "log": ["level": "error"],
            "dns": [
                "servers": [["type": "local", "tag": "dns-local"]],
                "final": "dns-local",
                "strategy": "prefer_ipv4",
            ] as [String: Any],
            "outbounds": outbounds,
            "route": ["final": "direct", "auto_detect_interface": false, "default_domain_resolver": "dns-local"] as [String: Any],
            "experimental": [
                "clash_api": ["external_controller": "127.0.0.1:\(settings.testerApiPort)", "secret": apiSecret],
            ],
        ]
        if !endpoints.isEmpty { config["endpoints"] = endpoints }
        return config
    }

    static func write(_ config: [String: Any], to url: URL) throws {
        let data = try JSONSerialization.data(withJSONObject: config, options: [.prettyPrinted, .sortedKeys])
        try data.write(to: url, options: .atomic)
    }
}
